src/client/java/daos-java/src/main/native/io_daos_obj_DaosObjClient.c

```diff
@@ -906,6 +1015,112 @@ Java_io_daos_obj_DaosObjClient_updateObjectSimple(
 	}
 }
 
+static int
+update_ret_code_no_decode(void *udata, daos_event_t *ev, int ret)
+{
+	data_desc_upd_async_t *desc = (data_desc_upd_async_t *)udata;
+	char *desc_buffer = desc->ret_buf_address;
+
+	memcpy(desc_buffer, &ret, 4);
+	if (ev) {
+		ev->ev_error = 0;
+	}
+	// free native desc if not reuse
+	if (desc->maxKeyLen == -1) {
+		release_desc_upd_async(desc);
+	}
+	return 0;
+}
+
+JNIEXPORT void JNICALL
+Java_io_daos_obj_DaosObjClient_updateObjNoDecode(JNIEnv *env,
+		jobject clientObj, jlong objPtr,
+		jlong descBufAddress, jlong eqWrapHdl,
+		jshort eqId, jlong offset, jint len,
+		jlong dataBufAddress)
+{
+	daos_handle_t oh;
+	char *desc_buf = (char *)descBufAddress;
+	char *data_buf = (char *)dataBufAddress;
+	data_desc_upd_async_t *desc = NULL;
+	uint64_t descPtr;
+	uint16_t value16;
+	daos_iod_t *iod;
+	int rc;
+
+	memcpy(&oh, &objPtr, 8);
+	memcpy(&descPtr, desc_buf, 8);
+	desc_buf += 8;
+	if (descPtr != 0L) { /* reusable */
+		/* skip maxlen */
+		desc_buf += 2;
+		desc = *(data_desc_upd_async_t **)&descPtr;
+		/* dkey */
+		memcpy(&value16, desc_buf, 2);
+		desc_buf += 2;
+		desc->dkey.iov_len = desc->dkey.iov_buf_len
+				= value16;
+		desc_buf += desc->maxKeyLen;
+		/* akey */
+		memcpy(&value16, desc_buf, 2);
+		desc_buf += 2;
+		iod = &desc->iods[0];
+		iod->iod_name.iov_len =
+			iod->iod_name.iov_buf_len = value16;
+		desc_buf += desc->maxKeyLen;
+	} else { /* not reusable */
+		char *msg = "memory allocation failed";
+		desc =	(data_desc_upd_async_t *)malloc(
+			sizeof(data_desc_upd_async_t));
+
+		if (desc == NULL) {
+			throw_const_obj(env,
+			msg,
+			CUSTOM_ERR3);
+			return;
+		}
+		rc = allocate_desc_update_async(desc_buf, desc, 0);
+		if (rc) {
+			throw_const_obj(env, "allocation failed",
+					rc);
+			return;
+		}
+
+	}
+	/* event */
+	event_queue_wrapper_t *eq = *(event_queue_wrapper_t **)
+				&eqWrapHdl;
+
+	desc->event = eq->events[eqId];
+	// offset
+	desc->recxs[0].rx_idx = offset;
+	/* length */
+	desc->recxs[0].rx_nr = (uint64_t)len;
+	/* sgl */
+	d_iov_set(&desc->iovs[0], data_buf, (uint64_t)len);
+	desc->sgls[0].sg_nr_out = 0;
+	rc = daos_event_register_comp_cb(desc->event,
+		update_ret_code_no_decode,
+		desc);
+	if (rc) {
+		char *msg = "Failed to register update callback";
+
+		throw_const_obj(env,
+				msg,
+				rc);
+		return;
+	}
+	// desc->event->ev_error = EVENT_IN_USE;
+	rc = daos_obj_update(oh, DAOS_TX_NONE, 0L, &desc->dkey,
+			1, desc->iods,
+			desc->sgls, desc->event);
+	if (rc) {
+		throw_const_obj(env,
+		 		"Failed to update DAOS object",
+		 		rc);
```

Comment:
(style)  code indent should use tabs where possible
(style)  please, no space before tabs